### DISCUSSION TITLE
Add Rooms to Vacuums

### DISCUSSION THREAD

**Discussion by Lash-L** [home-assistant/architecture#913] [Entity Models]
Per comments on [#93633](https://github.com/home-assistant/core/pull/93633):

A supported feature for rooms would be very beneficial. Being able to perform a specific action on a room in your house would be helpful, i.e. go to the living room or vacuum the kitchen.

I'm not sure what this would look like. 

Should it be a full map? like the [popular lovelace card?](https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card)

Should it be a selection/ multiselection on the Vacuum entity, you could select the room/ rooms you want to clean or select 'All' when you start the vacuum.

I'd be curious to hear other suggestions and I'd be happy to attempt to implement.

**Comment by allenporter** [comment]
I would suggest looking at a few different vacuums in order to ensure that your proposal can meet the needs of a few different devices -- the try to get craft a specific proposal or two.  You may want to propose the use cases you have in mind (e.g. similar to roborock app) then   what data and apis are needed for it.

My impression is that home assistant Areas are similar to rooms, perhaps that makes sense to use. Perhaps it is worth looking at areas closely to see if they make sense for this use case.    (I'm not sure if we today ever deal with devices that move between areas, so I think the device area should be static e.g. where the dock is.)

**Reply by balloob** [reply]
And to expand to this, maybe it would be super cool if we could sync the vacuum areas + floor plan somehow to HA…

**Reply by Lash-L** [reply]
Good point with the area system already built in to HA. 

And I agree with you, Paulus, that having a extremely accurate floor plan that can be used throughout your ha installation that is mapped for you by your robot would be extremely cool.

I'll write up a few detailed proposals tonight with some ideas of how we can do that and get back to you both

**Reply by Lash-L** [reply]
Started proposal for floor plan here: #950

**Comment by Lash-L** [comment]
for now - ignore and just look at https://github.com/home-assistant/architecture/discussions/913#discussioncomment-6600190

The tricky part of this whole task is to sync up rooms from a vacuum in a way that isn't super tedius.

To start off with - we would have three new support functionality(maybe just two to keep the first PR small)
1. supports area
2. supports clean area
3. supports go to area (maybe hold off on this on to keep inital PR smaller)

support areas would require you have a list of str of room names in the vacuum.

Now to actually connect the room to an area in Home Assistant:

Option 1:
This screen could have an additional Select or multi select drop down to select 'rooms' provided by an integration
![image](https://github.com/home-assistant/architecture/assets/20257911/ce531d6f-562c-4ace-a42a-a6acbd421030)
Perhaps you have multiple devices that have specific rooms inside of them, you can select all of them here. Maybe HA calls it Bedroom, Roborock calls it master bedroom and a speaker system calls it master. You could select Master Bedroom and master here and that way the area is linked to that room.

This only matters for devices that are in multiple rooms. Vacuums move not many other things do so I may be over complicating it. Most speakers are just one room, but theoretically if you have a receiver with multiple rooms, instead of making a new entity for each room, you can just have one entity, but add room support.

Option 2:
You could do something similar on the vacuum side, perhaps it shows all of the rooms from the vacuum and has a corresponding drop down where you can select what area it aligns with.


Now that you have rooms and areas synced up, you are able to call functions within the Vacuum such as clean room or go to room and it is synced with the corresponding area. So you would select the Area for cleaning and not the Roborock specific Room.

It gets slightly more complicated with chaining, i.e. if I want to clean 3 rooms, I should probably be able to do that rather than just performing one action at a time.

The Areas could add icons to them based on what devices have a 'room' synced. i.e. A clean icon or a music icon, so if you are looking directly at the area card, you can hit the clean button there.



As far as a floor plan goes, I think it makes sense for that to come after the original room support comes, but as a basic overview of how I would think that would work:

Probably a new entity, Map or floorplan or something similar, it has a collection of segments or rooms that have a name, are synced with an area, and have a polygon coordinate to describe where it is.

Floor plans would probably need to be device specific, as coordinates/ shape wouldn't match up very well. But each room should be a mini-area card or at least have the ability to add whatever buttons/ functionality you want. So you can easily perform actions on various rooms in your house.


I will continue to edit this as I get more feedback/ continue to think about it, the issue with the Switches earlier today set me back a little bit and I am a little brain dead. But I would be curious to hear thoughts on this so far.

**Comment by Lash-L** [comment]
I spent some time working on this the past few days, but there is a fundamental issue with connecting areas and rooms from devices that I was wondering if either of you had any ideas about.

Home Assistant Areas and Vacuum Rooms area can have a many to many relationship.

My vacuum could have two different rooms for kitchen, perhaps one for the kitchen area and one for the kitchen table.

Where as Home Assistant could have just one Area for the entire kitchen.

It could also work the opposite way. This makes things complicated, and I'm not sure if you have any good ideas on how to handle it.

The room support is built into the backend and I've done a rough start on the frontend.

![image](https://github.com/home-assistant/architecture/assets/20257911/01740d2c-58eb-461a-b08c-de6e63693451)
![image](https://github.com/home-assistant/architecture/assets/20257911/37c1ed6b-c0d8-4d08-a9fe-2fddcf096d63)
(Ignore the missing label for rooms)


However, a secondary issue is it doesn't seem like multiselect exist in the frontend like I thought it did. It exist for forms but not in general? Correct me if I am wrong. But that means this can only be used to clean one room at a time.

**Reply by allenporter** [reply]
Very nice exploration! I would suggest that it should only supporting 1:1 or 1:None mappings. Maybe separating these into two different concepts helps: (1) How do the vacuums and areas work together and (2) how does an integration (or floor plan) influence the set of areas being created.  For my own vacuum, I see cases where Roborock considers 3 areas one contiguous room:  In Home Assistant the areas are Kitchen, Living Room, Dining room.  I can a simple answer of just having the vacuum room name pick one of those areas (e.g. the Vacuum app has a single room name and it can just be one of the HA area names). For the other way where the vacuum has multiple rooms where Home Assistant has a single area, then HA would need to have additional areas created to control them, or it can't control them (maybe sub areas exist in the future, but it would probably need a lot more motivating cases beyond vacuums to introduce this concept)

Regarding multi-select there is a discussion just started: 
https://github.com/home-assistant/architecture/discussions/927

As you get more specific into the proposal, it would also be good to explore which other vacuum integrations / apis would be able to use this.

**Reply by shbatm** [reply]
> For my own vacuum, I see cases where Roborock considers 3 areas one contiguous room: In Home Assistant the areas are Kitchen, Living Room, Dining room.

Sidestepping a bit, but something I think is missing in Home Assistant areas which would help partiality address some of this is being able to nest Areas within other Areas (or create Groups of Areas). The requests have been lingering since 2019 [[1](https://community.home-assistant.io/t/nested-areas/125924), [2](https://community.home-assistant.io/t/why-the-heck-cant-areas-be-nested/224886)].

This is supported by Alexa (not sure about Google), but it's nice to be able to address areas at different levels of division.

It wouldn't solve this issue, but I think this should be incorporated as framework to help support things like these robot room mapping and the floor plan proposal. It would also be useful in Voice (e.g. Turn off Upstairs lights, which includes all upstairs areas).

**Comment by Lash-L** [comment]
> (1) How do the vacuums and areas work together

I see two main functionalities here:

1. If i say clean this Area, then the vacuum should clean that area, in my head if it were a many rooms: 1 area mapping, then that means the vacuum should clean each room that is contained within an area. If we allowed many Areas: 1 room, then if I say clean the kitchen area, but the kitchen area, living room area, and dining room area are all in one Roborock room, then I am actually cleaning all of those areas and doesn't reflect what I am looking for.
2. What area is my vacuum currently in? Again, if it were many rooms: 1 area, then if I am in any of the rooms within the area, then I am in that area.
3. Any others people can think of?



> (2) how does an integration (or floor plan) influence the set of areas being created

I think if it isn't helpful for the user, the ability to still clean rooms through a service or something would be rather helpful, that way if you don't care about having an Area for your living room hallway, but you want to be able to just clean your living room hallway, then you could call a service.

> >  I see cases where Roborock considers 3 areas one contiguous room: In Home Assistant the areas are Kitchen, Living Room, Dining room

I think the ideal solution to this is to change the rooms within the Roborock app to better correspond to the Area's in Home Assistant. If it is important for a user to be able to clean those three areas separately, then that should be reflected in the app and vice versa as well. If the user doesn't care, and when they clean their Home Assistant 'Kitchen', they are fine with the vacuum actually cleaning multiple rooms, then that would be okay as well. I think that again shows the strength of many rooms: 1 area.

> As you get more specific into the proposal, it would also be good to explore which other vacuum integrations / apis would be able to use this.

Agreed, I may reach out to a few other vacuum devs and see their thoughts.
